lib/client/db/postgres/repl/repl_test.go

```diff
@@ -0,0 +1,369 @@
+
+func TestStart(t *testing.T) {
+	ctx := context.Background()
+	_, tc := StartWithServer(t, ctx)
+
+	// Consume the REPL banner.
+	_ = readUntilNextLead(t, tc)
+
+	writeLine(t, tc, singleRowQuery)
+	singleRowQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "single", []byte(singleRowQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "single")), singleRowQueryResult)
+
+	writeLine(t, tc, multiRowQuery)
+	multiRowQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "multi", []byte(multiRowQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "multi")), multiRowQueryResult)
+
+	writeLine(t, tc, errorQuery)
+	errorQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "err", []byte(errorQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "err")), errorQueryResult)
+
+	writeLine(t, tc, dataTypesQuery)
+	dataTypeQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "data_type", []byte(dataTypeQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "data_type")), dataTypeQueryResult)
+}
+
+// TestQuery given some input lines, the REPL should execute the expected
+// query on the PostgreSQL test server.
+func TestQuery(t *testing.T) {
+	ctx := context.Background()
+	_, tc := StartWithServer(t, ctx, WithCustomQueries())
+
+	// Consume the REPL banner.
+	_ = readUntilNextLead(t, tc)
+
+	for name, tt := range map[string]struct {
+		lines         []string
+		expectedQuery string
+	}{
+		"query":                               {lines: []string{"SELECT 1;"}, expectedQuery: "SELECT 1;"},
+		"multiline query":                     {lines: []string{"SELECT", "1", ";"}, expectedQuery: "SELECT\r\n1\r\n;"},
+		"malformatted":                        {lines: []string{"SELECT err;"}, expectedQuery: "SELECT err;"},
+		"query with special characters":       {lines: []string{"SELECT 'special_chars_!@#$%^&*()';"}, expectedQuery: "SELECT 'special_chars_!@#$%^&*()';"},
+		"leading and trailing whitespace":     {lines: []string{"   SELECT 1;   "}, expectedQuery: "SELECT 1;"},
+		"multiline with excessive whitespace": {lines: []string{"   SELECT", "    1", "     ;"}, expectedQuery: "SELECT\r\n1\r\n;"},
+		// Commands should only be executed if they are at the beginning of the
+		// first line.
+		"with command in the middle":              {lines: []string{"SELECT \\d 1;"}, expectedQuery: "SELECT \\d 1;"},
+		"multiline with command in the middle":    {lines: []string{"SELECT", "\\d", ";"}, expectedQuery: "SELECT\r\n\\d\r\n;"},
+		"multiline with command in the last line": {lines: []string{"SELECT", "1", "\\d;"}, expectedQuery: "SELECT\r\n1\r\n\\d;"},
+	} {
+		t.Run(name, func(t *testing.T) {
+			for _, line := range tt.lines {
+				writeLine(t, tc, line)
+			}
+
+			require.EventuallyWithT(t, func(collect *assert.CollectT) {
+				query := <-tc.QueryChan()
+				require.Equal(t, tt.expectedQuery, query)
```

Comment:
similar to previous comment, feels a little weird to consume a channel in `require.EventuallyWithT`